torchtitan/experiments/flux/train.py

```diff
@@ -192,43 +213,308 @@ def train_step(self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor):
+                clip_encodings=empty_batch["clip_encodings"],
+                t5_encodings=empty_batch["t5_encodings"],
+                enable_classifer_free_guidance=self.job_config.eval.enable_classifer_free_guidance,
+                empty_t5_encodings=empty_batch["t5_encodings"],
+                empty_clip_encodings=empty_batch["clip_encodings"],
+                classifier_free_guidance_scale=self.job_config.eval.classifier_free_guidance_scale,
+            )
+
+        base_path = os.path.join(
+            self.job_config.job.dump_folder, self.job_config.eval.save_img_folder
         )
+        for i, image in enumerate(images):
+            name = (
+                f"image_rank{str(torch.distributed.get_rank())}_step{self.step}_{i}.png"
+            )
+            save_image(
+                name=name,
+                output_dir=base_path,
+                x=image,
+                prompt=inputs["txt"][i],
+            )
+        return images
+    
+    def generate_val_timesteps(self, cur_val_timestep, samples):
+        # generate timesteps for validation set
+        # first offset is the timesteps that have already been generated
+        # cycle through timesteps 0-7, repeating as necessary
+        first_offset = torch.arange(cur_val_timestep, 8, device=self.device)[:samples]
+        samples_left = samples - first_offset.numel()
+        val_timesteps = torch.arange(
+            0, 8, dtype=torch.int8, device=self.device
+        ).repeat_interleave(math.ceil(samples_left / 8))[:samples_left]
+        val_timesteps = torch.cat([first_offset, val_timesteps])
+        cur_val_timestep = (val_timesteps[-1].item() + 1) % 8
+        return val_timesteps, cur_val_timestep
+
+    @record
+    def train(self):
+        job_config = self.job_config
+
+        self.checkpointer.load(step=job_config.checkpoint.load_step)
+        logger.info(f"Training starts at step {self.step + 1}.")
+
+        with maybe_enable_profiling(
+            job_config, global_step=self.step
+        ) as torch_profiler, maybe_enable_memory_snapshot(
+            job_config, global_step=self.step
+        ) as memory_profiler, ft.maybe_semi_sync_training(
+            job_config,
+            ft_manager=self.ft_manager,
+            model=self.model_parts[0],
+            optimizer=self.optimizers,
+            sync_every=job_config.fault_tolerance.sync_steps,
+        ):
+            data_iterator = self.batch_generator(self.dataloader)
+            for inputs, labels in data_iterator:
+                if self.step >= job_config.training.steps:
+                    break
+                self.step += 1
+                self.gc_handler.run(self.step)
+                self.train_step(inputs, labels)
+                self.checkpointer.save(
+                    self.step, force=(self.step == job_config.training.steps)
+                )
+
+                if self.step % job_config.eval.eval_freq == 0 and job_config.eval.dataset:
```

Comment:
I think we could wrap all these parts into eval(), and make the main train() loop easier to read.